prime-router/src/main/kotlin/fhirengine/translation/hl7/ConfigSchemaProcessor.kt

```diff
@@ -0,0 +1,136 @@
+package gov.cdc.prime.router.fhirengine.translation.hl7
+
+import gov.cdc.prime.router.fhirengine.translation.hl7.schema.ConfigSchemaElement
+import gov.cdc.prime.router.fhirengine.translation.hl7.utils.CustomContext
+import gov.cdc.prime.router.fhirengine.translation.hl7.utils.FhirPathUtils
+import org.apache.logging.log4j.kotlin.Logging
+import org.hl7.fhir.r4.model.Base
+import org.hl7.fhir.r4.model.Bundle
+
+abstract class ConfigSchemaProcessor : Logging {
+
+    /**
+     * Get the first valid string from the list of values specified in the schema for a given [element] using
+     * [bundle] and [context] starting at the [focusResource].
+     * @return the value for the element or an empty string if no value found
+     */
+    internal fun getValueAsString(
+        element: ConfigSchemaElement,
+        bundle: Bundle,
+        focusResource: Base,
+        context: CustomContext
+    ): String {
+        var retVal = ""
+        run findValue@{
+            element.value.forEach {
+                val value = if (it.isBlank()) ""
+                else try {
+                    FhirPathUtils.evaluateString(context, focusResource, bundle, it)
+                } catch (e: SchemaException) {
+                    logger.error("Error while getting value for element ${element.name}", e)
+                    ""
+                }
+                logger.trace("Evaluated value expression '$it' to '$value'")
+                if (value.isNotBlank()) {
+                    retVal = value
+                    return@findValue
+                }
+            }
+        }
+
+        // when valueSet is available, use the matching value else just pass the value as is
+        // does a lowerCase comparison
+        if (element.valueSet.isNotEmpty()) {
+            val lowerSet = element.valueSet.mapKeys { it.key.lowercase() }
+            retVal = lowerSet.getOrDefault(retVal.lowercase(), retVal)
+        }
+        return retVal
+    }
+
+    /**
+     * Get the first valid value from the list of values specified in the schema for a given [element] using
+     * [bundle] and [context] starting at the [focusResource].
+     * @return the value for the element or null if no value found
+     */
+    internal fun getValue(
+        element: ConfigSchemaElement,
+        bundle: Bundle,
+        focusResource: Base,
+        context: CustomContext
+    ): Base? {
+        var retVal: Base? = null
+        run findValue@{
+            element.value.forEach {
+                val value = if (it.isBlank()) emptyList<Base>()
+                else try {
+                    FhirPathUtils.evaluate(context, focusResource, bundle, it)
+                } catch (e: SchemaException) {
+                    logger.error("Error while getting value for element ${element.name}", e)
+                    emptyList<Base>()
```

Comment:
Missing code coverage

Reply:
Didn't realize until looking into this more, but the `evaluate`, unlike `evaluateString`, doesn't throw a schema exception that I can find, so I'm just removing the try/catch here.